lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,119 @@
+"""
+Middleware for Mobile APIs
+"""
+from datetime import datetime
+from django.core.cache import cache
+from django.http import HttpResponse
+from pytz import UTC
+from mobile_api.mobile_platform import MobilePlatform
+from mobile_api.models import AppVersionConfig
+from mobile_api.utils import parsed_version
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+from request_cache.middleware import RequestCache
+
+
+class AppVersionUpgrade(object):
+    """
+    Middleware class to keep track of mobile application version being used
+    """
+    LATEST_VERSION_HEADER = 'EDX-APP-LATEST-VERSION'
+    UPGRADE_DEADLINE_HEADER = 'EDX-APP-UPGRADE-DATE'
+    NO_LAST_SUPPORTED_DATE = 'NO_LAST_SUPPORTED_DATE'
+    NO_LATEST_VERSION = 'NO_LATEST_VERSION'
+
+    def process_request(self, request):
+        """
+        raises HTTP Upgrade Require error if request is from mobile native app and
+        user app version is no longer supported
+        """
+        user_agent = request.META.get('HTTP_USER_AGENT')
+        request_cache_dict = self.set_request_cache(request, user_agent)
+        last_supported_date = (request_cache_dict[self.UPGRADE_DEADLINE_HEADER]
+                            if self.UPGRADE_DEADLINE_HEADER in request_cache_dict else None)
+        if last_supported_date and last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+            if datetime.now().replace(tzinfo=UTC) > last_supported_date:
+                return HttpResponse(status=426)
+
+    def process_response(self, _request, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < deadline to upgrade
+        """
+        request_cache_dict = RequestCache.get_request_cache().data
```

Comment:
The request cache access is spread around in too methods. If we rename `set_request_cache()` to `get_version_info(request)`, have it deal with the request cache and return a dictionary which always has two keys `latest_version` and `last_supported`, `process_request()` and `process_response()` logic will become much clearer.